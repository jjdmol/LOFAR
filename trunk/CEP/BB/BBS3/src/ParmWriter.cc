//#  ParmWriter.cc: writes/updates parameters in ParmTables
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <BBS3/ParmWriter.h>
#include <BBS3/ParmData.h>
#include <MNS/MeqPolc.h>
#include <MNS/MeqDomain.h>
#include <MNS/ParmTable.h>
#include <Common/LofarLogger.h>

#include <BBS3/MNS/MeqStoredParmPolc.h>

using namespace std;

namespace LOFAR
{
ParmWriter::ParmWriter(const int DBMasterPort)
  : itsDBMasterPort(DBMasterPort)
{}

ParmWriter::~ParmWriter()
{}

void ParmWriter::write(vector<ParmData>& pData, double fStart, double fEnd,
		       double tStart, double tEnd)
{
  MeqDomain domain(fStart, fEnd, tStart, tEnd);

  // Store all parameters in their own ParmTable
  // Use one ParmTable at the time to avoid reopening the same table.
  vector<char> done(pData.size(), 0); 
  MeqParmGroup pgroup;
  uint nrDone = 0;
  string lastName;
  ParmDB::ParmDB* dbptr = 0;
  while (nrDone < pData.size()) {
    for (uint i=0; i<pData.size(); ++i) {
      if (done[i] == 0) {
	if (dbptr == 0) {
	  lastName = pData[i].getParmDBMeta().getTableName();
	  dbptr = new ParmDB::ParmDB(pData[i].getParmDBMeta());
	}
	if (pData[i].getParmDBMeta().getTableName() == lastName) {
//     cout << "Writing parm " << pData[i].getName() 
//  	 << " values=" << pData[i].getValues() << endl;
	  MeqStoredParmPolc parm(pData[i].getName(), &pgroup, dbptr);
	  parm.readPolcs (domain);
	  parm.update (pData[i].getValues());
	  parm.save();
	  done[i] = 1;
	  nrDone++;
//     streamsize prec = cout.precision();
//     cout.precision(10);
//     cout << "****Contr write: " << parm.getCoeffValues().getDouble()
// 	 << " for parameter " << pData[i].getName() << endl;
//     cout.precision (prec);
	}
      }
    }
    delete dbptr;
    dbptr = 0;
  }
}

} // namespace LOFAR
