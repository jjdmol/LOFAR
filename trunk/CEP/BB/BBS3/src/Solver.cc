//# Solver.cc: Calculate parameter values using a least squares fitter
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#include <lofar_config.h>

#include <BBS3/Solver.h>
#include <Common/VectorUtil.h>
#include <Common/LofarLogger.h>
#include <Common/Timer.h>

#include <iostream>
#include <fstream>
#include <stdexcept>


namespace LOFAR
{

//----------------------------------------------------------------------
//
// ~Solver
//
// Constructor. Initialize a Solver object.
//
// Create list of stations and list of baselines from MS.
// Create the MeqExpr tree for the WSRT.
//
//----------------------------------------------------------------------
Solver::Solver ()
: itsSolver (1),
  itsDoSet  (true)
{
  LOG_INFO_STR( "Solver constructor" );
}

//----------------------------------------------------------------------
//
// ~~Solver
//
// Destructor for a Solver object.
//
//----------------------------------------------------------------------
Solver::~Solver()
{
  LOG_TRACE_FLOW( "Solver destructor" );
}


// ~solve
//
// Solve for the solvable parameters on the current time domain.
// Returns solution and fitness.
//
//----------------------------------------------------------------------

void Solver::solve (bool useSVD,
		    Quality& resultQuality)
{
  LOG_INFO_STR( "solve using file ");

  NSTimer timer;
  timer.start();
  ASSERT (itsSolvableValues.size() > 0);

  // Initialize the solver (needed after a setSolvable).
  // Note it is usually already done in setEquations.
  if (itsDoSet) {
    itsSolver.set (itsSolvableValues.size());
    itsDoSet = false;
  }
  // Solve the equation. 
  uint rank;
  double fit;
  LOG_INFO_STR( "Solution before: " << itsSolvableValues);
  bool solFlag = itsSolver.solveLoop (fit, rank, &(itsSolvableValues[0]),
				      useSVD);
  LOG_INFO_STR( "Solution after:  " << itsSolvableValues);

  resultQuality.init();
  resultQuality.itsSolFlag = solFlag;
  resultQuality.itsRank = rank;
  resultQuality.itsFit = fit;
  resultQuality.itsMu = itsSolver.getWeightedSD();
  resultQuality.itsStddev = itsSolver.getSD();
  resultQuality.itsChi = itsSolver.getChi();
  cout << resultQuality << endl;

  // Store the new values in the ParmData vector.
  const double* val = &(itsSolvableValues[0]);
  for (uint i=0; i<itsSolvableParms.size(); ++i) {
    ParmData& parm = itsSolvableParms[i];
    double* value = parm.getRWValues().doubleStorage();
    for (int j=0; j<parm.getNrSpid(); ++j) {
      value[j] = *val++;
    }
  }

  timer.stop();
  cout << "BBSTest: solver     " << timer << endl;
  return;
}


void Solver::setEquations (const double* data, int nresult, int nrspid,
			   int nval, int prediffer)
{
  ASSERT (uint(prediffer) < itsIndices.size());
  // Initialize the solver (needed after a setSolvable).
  if (itsDoSet) {
    itsSolver.set (itsSolvableValues.size());
    itsDoSet = false;
  }
  vector<int>& predInx = itsIndices[prediffer];
  ASSERT (uint(nrspid) == predInx.size());
  // Use a consecutive vector to assemble all derivatives.
  vector<double> derivVec(nrspid);
  double* derivs = &(derivVec[0]);
  // Each result is a 2d array of [nrval,nrspid+1] (nrval varies most rapidly).
  // The first value is the difference; the others the derivatives. 
  const double* ddata = data;
  int nrval = nval;
  for (int i=0; i<nresult; ++i) {
    for (int j=0; j<nrval; ++j) {
      // Each value result,time,freq gives an equation.
      double diff = ddata[0];
      const double* derivdata = ddata + nrval;
      for (int k=0; k<nrspid; ++k) {
	derivs[k] = derivdata[k*nrval];
      }
      itsSolver.makeNorm (predInx.size(), &(predInx[0]), &(derivVec[0]),
      			  1., diff);
      // Go to next time,freq value.
      ddata++;
    }
    // Go to next result (note that ddata has already been increment nrval
    // time, so here we use nrspid instead of nrspid+1.
    ddata += nrspid*nrval;
  }
}

void Solver::initSolvableParmData (int nrPrediffers)
{
  itsSolvableMap.clear();
  itsSolvableParms.resize (0);
  itsIndices.resize (nrPrediffers);
  itsSolvableValues.resize (0);
  itsDoSet = true;
}

// The solvable parameters coming from the prediffers have to be collected
// and put into a single solvercollection.
// It means that the prediffer spids have to be mapped to a solver spids.
void Solver::setSolvableParmData (const std::vector<ParmData>& data,
				  int prediffer)
{
  if (data.size() == 0) {
    return;
  }
  itsDoSet = true;
  // Usually the last entry has the highest spid; so use that to reserve to
  // avoid (hopefully) resizes.
  vector<int> predInx;
  predInx.reserve (data[data.size()-1].getLastSpid() + 1);
  // Loop through all parm entries and see if already used.
  for (uint i=0; i<data.size(); ++i) {
    int parminx = itsSolvableMap.size();
    int spidinx = itsSolvableValues.size();
    ParmData parm(data[i]);
    map<string,int>::const_iterator value =
                                     itsSolvableMap.find (parm.getName());
    if (value == itsSolvableMap.end()) {
      // Not in use yet; so add it to the map.
      itsSolvableMap[parm.getName()] = parminx;
      // Its first solver spid is the length of the vector so far.
      parm.setFirstSpid (spidinx);
      // Save the parm.
      itsSolvableParms.push_back (parm);
      // Store the current values in the vector.
      const double* value = parm.getValues().doubleStorage();
      for (int i=0; i<parm.getNrSpid(); ++i) {
	itsSolvableValues.push_back (value[i]);
      }
    } else {
      // The parameter has already been used. Get its index.
      parminx = value->second;
      ASSERT (itsSolvableParms[parminx] == parm);
      spidinx = itsSolvableParms[parminx].getFirstSpid();
    }
    // Store the solver spid to get a mapping of prediffer spid to solver spid.
    int spidnr = parm.getFirstSpid();
    for (int i=0; i<parm.getNrSpid(); ++i) {
      predInx.push_back (spidnr++);
    }
  }
  ASSERT (uint(prediffer) < itsIndices.size());
  itsIndices[prediffer] = predInx;
}

} // namespace LOFAR

//# Instantiate the makeNorm template.
#include <scimath/Fitting/LSQFit2.cc>
template void casa::LSQFit::makeNorm<double, double*, int*>(unsigned, int* const&, double* const&, double const&, double const&, bool, bool);
