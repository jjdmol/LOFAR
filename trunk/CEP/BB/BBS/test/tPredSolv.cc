//# tPredSolv.cc: Test program for Prediffer and Solver classes
//#
//# Copyright (C) 2005
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#include <lofar_config.h>
#include <BBS/Prediffer.h>
#include <BBS/Solver.h>
#include <BBS/ParmWriter.h>
#include <BBS/MNS/MeqParmFunklet.h>
#include <ParmDB/ParmDB.h>
#include <Common/VectorUtil.h>
#include <Common/LofarLogger.h>
#include <stdexcept>
#include <iostream>
#include <iomanip>
#include <malloc.h>

#include <Common/hexdump.h>


using namespace LOFAR;
using namespace std;

// Note:
// demo3.MS contains 50 frequency channels of 500000 Hz with
// center frequencies of 137750000-162250000 Hz.
// There are 5 time stamps of 2 sec in it (centers 2.35208883e9 + 2-10).

void doSolve (Prediffer& pre1, const vector<string>& solv, int niter)
{
  // Set the solvable parameters.
  pre1.clearSolvableParms();
  pre1.setSolvableParms (solv, vector<string>());
  // Set a domain.
  cout << ">>>" << endl;
  pre1.setWorkDomain (0, 1, 0., 1e12);
  //  pre1.setWorkDomain (137750000-250000, 2*500000, 0., 1e12);
  //  pre1.setWorkDomain (0., 1e12, 0., 1e12);
  vector<MeqDomain> solveDomains(1, pre1.getWorkDomain());
  pre1.initSolvableParms (solveDomains);
  cout << "<<<" << endl;
  // Get the ParmData from the Prediffer and send it to the solver.
  Solver solver;
  solver.initSolvableParmData (1, solveDomains, pre1.getWorkDomain());
  solver.setSolvableParmData (pre1.getSolvableParmData(), 0);
  cout << ">>>" << endl;
  pre1.showSettings();
  cout << "<<<" << endl;
  streamsize prec = cout.precision();
  cout << "Before: " << setprecision(10) << solver.getSolvableValues(0)
       << endl;

  for (int it=0; it<niter; ++it) {
    // Get the fitter from the prediffer and give it to the solver.
    vector<casa::LSQFit> fitters;
    cout << ">>>" << endl;
    pre1.fillFitters (fitters, "DATA");
    solver.mergeFitters (fitters, 0);
    // Do the solve.
    solver.solve (false);
    cout << "<<<" << endl;
    cout << "iter" << it << ":  " << setprecision(10)
	 << solver.getSolvableValues(0) << endl;
    cout << solver.getQuality(0) << endl;
    cout.precision (prec);
    pre1.updateSolvableParms (solver.getSolvableParmData());
  }
}

void doSolve2 (Prediffer& pre1, Prediffer& pre2,
	       const vector<string>& solv, int niter)
{
  // Set the solvable parameters.
  pre1.clearSolvableParms();
  pre2.clearSolvableParms();
  pre1.setSolvableParms (solv, vector<string>());
  pre2.setSolvableParms (solv, vector<string>());
  // Set a domain.
  cout << ">>>" << endl;
  pre1.setWorkDomain (0, 3, 0., 1e12);
  pre2.setWorkDomain (0, 3, 0., 1e12);
  vector<MeqDomain> solveDomains(1, pre1.getWorkDomain());
  pre1.initSolvableParms (solveDomains);
  pre2.initSolvableParms (solveDomains);
  cout << "<<<" << endl;
  // Get the ParmData from the Prediffers and send it to the solver.
  Solver solver;
  solver.initSolvableParmData (2, solveDomains, pre1.getWorkDomain());
  solver.setSolvableParmData (pre1.getSolvableParmData(), 0);
  solver.setSolvableParmData (pre2.getSolvableParmData(), 1);
    
  streamsize prec = cout.precision();
  cout << "Before: " << setprecision(10) << solver.getSolvableValues(0)
       << endl;
  for (int it=0; it<niter; ++it) {
    // Get the fitter from the prediffer and give it to the solver.
    vector<casa::LSQFit> fitters;
    cout << ">>>" << endl;
    pre1.fillFitters (fitters, "DATA");
    solver.mergeFitters (fitters, 0);
    pre2.fillFitters (fitters, "DATA");
    solver.mergeFitters (fitters, 1);
    // Do the solve.
    solver.solve (false);
    cout << "<<<" << endl;
    cout << "iter" << it << ":  " << setprecision(10)
	 << solver.getSolvableValues(0) << endl;
    cout << solver.getQuality(0) << endl;
    cout.precision (prec);
    pre1.updateSolvableParms (solver.getSolvableParmData());
    pre2.updateSolvableParms (solver.getSolvableParmData());
  }
}

void doSolve1 (Prediffer& pre1, const vector<string>& solv, int niter)
{
  // Set the solvable parameters.
  pre1.clearSolvableParms();
  pre1.setSolvableParms (solv, vector<string>());
  // Set a domain.
  cout << ">>>" << endl;
  pre1.setWorkDomain (0, 3, 0., 1e12);
  //pre1.setWorkDomain (0., 1e12, 0., 1e12);
  vector<MeqDomain> solveDomains(1, pre1.getWorkDomain());
  pre1.initSolvableParms (solveDomains);
  cout << "<<<" << endl;
  // Get the ParmData from the Prediffer and send it to the solver.
  Solver solver;
  solver.initSolvableParmData (1, solveDomains, pre1.getWorkDomain());
  const ParmDataInfo& pData = pre1.getSolvableParmData();
  solver.setSolvableParmData (pData, 0);
  cout << ">>>" << endl;
  pre1.showSettings();
  cout << "<<<" << endl;
  streamsize prec = cout.precision();
  cout << "Before: " << setprecision(10) << solver.getSolvableValues(0)
       << endl;

  for (int it=0; it<niter; ++it) {
    // Get the fitter from the prediffer and give it to the solver.
    vector<casa::LSQFit> fitters;
    cout << ">>>" << endl;
    pre1.fillFitters (fitters, "DATA");
    solver.mergeFitters (fitters, 0);
    // Do the solve.
    solver.solve (false);
    cout << "<<<" << endl;
    cout << "iter" << it << ":  " << setprecision(10)
	 << solver.getSolvableValues(0) << endl;
    cout << solver.getQuality(0) << endl;
    cout.precision (prec);
    MeqDomain domain(pre1.getWorkDomain());
    ParmWriter pwriter;
    pwriter.write (solver.getSolvableParmData(),
		   domain.startX(), domain.endX(),
		   domain.startY(), domain.endY());
    pre1.updateSolvableParms();
  }
}


int main (int argc, const char* argv[])
{
  INIT_LOGGER("tPredSolv");
  try {
    if (argc < 5) {
      cerr << "Run as: tPredSolv user msname meqparmtable skyparmtable"
	   << endl;
      return 1;
    }

    // Read the info for the ParmTables
    ParmDB::ParmDBMeta meqPdm("aips", argv[3]);
    ParmDB::ParmDBMeta skyPdm("aips", argv[4]);

    // Do a solve for RA using a few stations.
    {
      cout << "Starting first test" << endl;
      vector<int> antVec(10);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = 2*i;
      }
      vector<vector<int> > srcgrp;
      Prediffer pre1(argv[2], meqPdm, skyPdm, 
		     antVec, "TOTALEJ.REALIMAG", srcgrp, false);
      // Do a further selection of a few stations.
      vector<int> antVec2(10);
      for (uint i=0; i<antVec2.size(); ++i) {
	antVec2[i] = 4*i;
      }
      vector<int> corr;
      pre1.select (antVec2, antVec2, false, corr);    // no autocorrelations
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve (pre1, solv, 9);
      cout << "End of first test" << endl;
    }
    // Do a solve using 2 prediffers.
    {
      cout << "Starting test with two prediffers" << endl;
      vector<int> antVec(10);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = 2*i;
      }
      vector<vector<int> > srcgrp;
      Prediffer pre1(argv[2], meqPdm, skyPdm, 
		     antVec, "TOTALEJ.REALIMAG", srcgrp, false);
      Prediffer pre2(argv[2], meqPdm, skyPdm, 
		     antVec, "TOTALEJ.REALIMAG", srcgrp, false);
      // Do a further selection of a few stations.
      vector<int> antVec2(10);
      for (uint i=0; i<antVec2.size(); ++i) {
	antVec2[i] = 4*i;
      }
      vector<int> corr;
      pre1.select (antVec2, antVec2, false, corr);    // no autocorrelations
      pre2.select (antVec2, antVec2, false, corr);    // no autocorrelations
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve2 (pre1, pre2, solv, 1);
      cout << "End of test with two prediffers" << endl;
    }
    // Take more baselines.
    {
      cout << "Starting test with 21 antennas" << endl;
      vector<int> antVec(21);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = 4*i;
      }
      vector<vector<int> > srcgrp;
      Prediffer pre1(argv[2], meqPdm, skyPdm, 
		     antVec, "TOTALEJ.REALIMAG", srcgrp, false);
      // Only use first correlation.
      vector<int> corrVec(1, 0);
      vector<int> antVec2;
      pre1.select (antVec2, antVec2, false, corrVec);    // no autocorrelations
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve (pre1, solv, 5);
      cout << "End of test with 21 antennas" << endl;
    }
    // Do a solve updating the parm table.
    // This should be the last one.
    {
      cout << "Starting test with updating parmtable" << endl;
      vector<int> antVec(10);
      for (uint i=0; i<antVec.size(); ++i) {
	antVec[i] = 2*i;
      }
      vector<vector<int> > srcgrp;
      Prediffer pre1(argv[2], meqPdm, skyPdm, 
		     antVec, "TOTALEJ.REALIMAG", srcgrp, false);
      // Do a further selection of a few stations.
      vector<int> antVec2(10);
      for (uint i=0; i<antVec2.size(); ++i) {
	antVec2[i] = 4*i;
      }
      vector<int> corr;
      pre1.select (antVec2, antVec2, false, corr);    // no autocorrelations
      vector<string> solv(3);
      solv[0] = "RA.*";
      solv[1] = "DEC.*";
      solv[2] = "StokesI.*";
      doSolve1 (pre1, solv, 5);
      cout << "End of test with updating parmtable" << endl;
    }

  } catch (std::exception& x) {
    cerr << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
