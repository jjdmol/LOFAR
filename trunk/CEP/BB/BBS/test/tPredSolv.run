#!/bin/sh

# Set srcdir if needed.
# It might not be set if the .run file is executed directly instead of
# by 'make check'.
if [ "$srcdir" = "" ]; then
  srcdir=../../../test
fi

# Create the ParmDB tables for tPredSolv.
$srcdir/makeDemoParmDB tPredSolv

$LOFAR_CHECKTOOL ./tPredSolv $USER /opt/lofar/data/BBS2/demo3.MS_p1 tPredSolv_tmp.demo_instr tPredSolv_tmp.demo_sky
status=$?
exit $status

