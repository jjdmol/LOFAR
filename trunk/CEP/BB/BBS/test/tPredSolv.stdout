creating instrument_model (aips++)...
creating sky_model (aips++)...
Starting first test
>>>
<<<
>>>
Prediffer settings:
  msname:    /data/PSS3/diepen/BBS2/demo3.MS_p1
  mepname:   tPredSolv_tmp.demo_instr
  gsmname:   tPredSolv_tmp.demo_sky
  solvparms: [DEC:SRC1(0),DEC:SRC2(0),DEC:SRC3(0),RA:SRC1(0),RA:SRC2(0),RA:SRC3(0),StokesI:SRC1(0),StokesI:SRC2(0),StokesI:SRC3(0)]
  stchan:    0
  endchan:   1
  corr     : 4  1,1,1,1
  calcuvw  : 0

<<<
Before: [0.4537895,0.45368975,0.45375025,2.734003,2.7340205,2.733981,1,1,1]
>>>
<<<
iter0:  [0.4538001731,0.4536930993,0.4537583416,2.734004076,2.734020405,2.733981394,0.6039205533,0.5974054979,0.6004018856]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.01102832013
itsStddev : 0.01102832013
itsChi    : 0.09620446133

>>>
<<<
iter1:  [0.4538008139,0.4536965908,0.4537684625,2.734004501,2.734017502,2.733984812,0.6056732548,0.5944160394,0.599921246]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.997991457
itsMu     : 0.0001383415025
itsStddev : 0.0001383415025
itsChi    : 1.513845172e-05

>>>
<<<
iter2:  [0.4537966951,0.4536953623,0.4537734395,2.734005357,2.734016631,2.733984825,0.6076239878,0.5926366283,0.5997288336]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.998236256
itsMu     : 4.83382342e-05
itsStddev : 4.83382342e-05
itsChi    : 1.848238645e-06

>>>
<<<
iter3:  [0.4537966951,0.4536953623,0.4537734395,2.734005357,2.734016631,2.733984825,0.6076239878,0.5926366283,0.5997288336]
itsSolFlag: 0
itsRank   : 0
itsFit    : -1.64277398
itsMu     : 0
itsStddev : 0
itsChi    : 0

>>>
<<<
iter4:  [0.4537964747,0.4536955569,0.4537734839,2.734005349,2.73401662,2.733984837,0.6076499642,0.5925934234,0.5997525705]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 4.455541869e-05
itsStddev : 4.455541869e-05
itsChi    : 3.158429867e-06

>>>
<<<
iter5:  [0.4537960501,0.453695427,0.4537740044,2.734005506,2.734016505,2.733984796,0.6078078423,0.5924074286,0.5997745074]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.562235683
itsMu     : 4.301307676e-05
itsStddev : 4.301307676e-05
itsChi    : 1.463448695e-06

>>>
<<<
iter6:  [0.4537954918,0.4536952294,0.4537746398,2.73400571,2.734016377,2.733984717,0.6087007697,0.591780055,0.5995094074]
itsSolFlag: 1
itsRank   : 9
itsFit    : -0.06602045315
itsMu     : 4.174575154e-05
itsStddev : 4.174575154e-05
itsChi    : 1.378481848e-06

>>>
<<<
iter7:  [0.4537954918,0.4536952294,0.4537746398,2.73400571,2.734016377,2.733984717,0.6087007697,0.591780055,0.5995094074]
itsSolFlag: 0
itsRank   : 0
itsFit    : -0.04196550248
itsMu     : 0
itsStddev : 0
itsChi    : 0

>>>
<<<
iter8:  [0.4537954822,0.4536952313,0.4537746473,2.734005712,2.734016374,2.733984717,0.6087097309,0.5917732443,0.599507343]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 4.145217827e-05
itsStddev : 4.145217827e-05
itsChi    : 2.733788385e-06

End of first test
Starting test with two prediffers
>>>
<<<
Before: [0.4537895,0.45368975,0.45375025,2.734003,2.7340205,2.733981,1,1,1]
>>>
<<<
iter0:  [0.4538001633,0.4536931018,0.4537583473,2.734004074,2.734020395,2.733981406,0.6039471597,0.5973809337,0.6004009831]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.01099723109
itsStddev : 0.01099723109
itsChi    : 0.1924140947

End of test with two prediffers
Starting test with 21 antennas
>>>
<<<
>>>
Prediffer settings:
  msname:    /data/PSS3/diepen/BBS2/demo3.MS_p1
  mepname:   tPredSolv_tmp.demo_instr
  gsmname:   tPredSolv_tmp.demo_sky
  solvparms: [DEC:SRC1(0),DEC:SRC2(0),DEC:SRC3(0),RA:SRC1(0),RA:SRC2(0),RA:SRC3(0),StokesI:SRC1(0),StokesI:SRC2(0),StokesI:SRC3(0)]
  stchan:    0
  endchan:   1
  corr     : 4  1,0,0,0
  calcuvw  : 0

<<<
Before: [0.4537895,0.45368975,0.45375025,2.734003,2.7340205,2.733981,1,1,1]
>>>
<<<
iter0:  [0.4537900035,0.4536898874,0.4537501573,2.734000058,2.734020269,2.733980649,0.9838712296,0.5025455826,0.3049662164]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.02657902437
itsStddev : 0.02657902437
itsChi    : 2.960709051

>>>
<<<
iter1:  [0.4537900019,0.4536900032,0.4537499987,2.734000002,2.734020009,2.733980009,1.000212808,0.5000044708,0.2993235781]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.998141673
itsMu     : 0.0007906191819
itsStddev : 0.0007906191819
itsChi    : 0.002619704793

>>>
<<<
iter2:  [0.45379,0.45369,0.4537499998,2.734,2.73402,2.73398,0.9999948478,0.49998994,0.3000014784]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.993345788
itsMu     : 1.492073604e-05
itsStddev : 1.492073604e-05
itsChi    : 9.330354729e-07

>>>
<<<
iter3:  [0.45379,0.45369,0.45375,2.734,2.73402,2.73398,0.9999998111,0.4999999478,0.3000003025]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.998783353
itsMu     : 2.789021435e-07
itsStddev : 2.789021435e-07
itsChi    : 3.260028261e-10

>>>
<<<
iter4:  [0.45379,0.45369,0.45375,2.734,2.73402,2.73398,1.000000002,0.5000000021,0.2999999998]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.997468976
itsMu     : 2.30580912e-08
itsStddev : 2.30580912e-08
itsChi    : 2.228252312e-12

End of test with 21 antennas
Starting test with updating parmtable
>>>
<<<
>>>
Prediffer settings:
  msname:    /data/PSS3/diepen/BBS2/demo3.MS_p1
  mepname:   tPredSolv_tmp.demo_instr
  gsmname:   tPredSolv_tmp.demo_sky
  solvparms: [DEC:SRC1(0),DEC:SRC2(0),DEC:SRC3(0),RA:SRC1(0),RA:SRC2(0),RA:SRC3(0),StokesI:SRC1(0),StokesI:SRC2(0),StokesI:SRC3(0)]
  stchan:    0
  endchan:   3
  corr     : 4  1,1,1,1
  calcuvw  : 0

<<<
Before: [0.4537895,0.45368975,0.45375025,2.734003,2.7340205,2.733981,1,1,1]
>>>
<<<
iter0:  [0.4538001633,0.4536931018,0.4537583473,2.734004074,2.734020395,2.733981406,0.6039471597,0.5973809337,0.6004009831]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 0.01099723109
itsStddev : 0.01099723109
itsChi    : 0.1924140947

>>>
<<<
iter1:  [0.4538007895,0.4536965748,0.4537684948,2.734004511,2.734017495,2.733984811,0.6057240058,0.5943667748,0.5999195407]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.997977332
itsMu     : 0.0001387250705
itsStddev : 0.0001387250705
itsChi    : 3.061823049e-05

>>>
<<<
iter2:  [0.4537966814,0.4536953566,0.4537734474,2.734005359,2.734016629,2.733984825,0.6077026266,0.5925676336,0.5997190639]
itsSolFlag: 1
itsRank   : 9
itsFit    : -1.998215963
itsMu     : 4.860151325e-05
itsStddev : 4.860151325e-05
itsChi    : 3.758112381e-06

>>>
<<<
iter3:  [0.4537966814,0.4536953566,0.4537734474,2.734005359,2.734016629,2.733984825,0.6077026266,0.5925676336,0.5997190639]
itsSolFlag: 0
itsRank   : 0
itsFit    : -1.647615655
itsMu     : 0
itsStddev : 0
itsChi    : 0

>>>
<<<
iter4:  [0.4537964622,0.4536955494,0.4537734924,2.734005351,2.734016618,2.733984837,0.6077289227,0.5925240488,0.5997427985]
itsSolFlag: 1
itsRank   : 9
itsFit    : 1
itsMu     : 4.491003141e-05
itsStddev : 4.491003141e-05
itsChi    : 6.435962751e-06

End of test with updating parmtable
